src/commands/close.rs

```diff
@@ -0,0 +1,163 @@
+};
+
+#[derive(Debug, clap::Parser)]
+pub struct CloseOptions {
+    /// Close Pull Requests for the whole branch, not just the HEAD commit
+    #[clap(long)]
+    all: bool,
+}
+
+pub async fn close(
+    opts: CloseOptions,
+    git: &crate::git::Git,
+    gh: &mut crate::github::GitHub,
+    config: &crate::config::Config,
+) -> Result<()> {
+    let mut result = Ok(());
+
+    let mut prepared_commits = git.get_prepared_commits(config)?;
+
+    if prepared_commits.is_empty() {
+        output("👋", "Branch is empty - nothing to do. Good bye!")?;
+        return result;
+    };
+
+    if !opts.all {
+        // Remove all prepared commits from the vector but the last. So, if
+        // `--all` is not given, we only operate on the HEAD commit.
+        prepared_commits.drain(0..prepared_commits.len() - 1);
+    }
+
+    for prepared_commit in prepared_commits.iter_mut() {
+        if result.is_err() {
+            break;
+        }
+
+        write_commit_title(prepared_commit)?;
+
+        // The further implementation of the close command is in a separate function.
+        // This makes it easier to run the code to update the local commit message
+        // with all the changes that the implementation makes at the end, even if
+        // the implementation encounters an error or exits early.
+        result = close_impl(gh, config, prepared_commit).await;
+    }
+
+    // This updates the commit message in the local Git repository (if it was
+    // changed by the implementation)
+    add_error(
+        &mut result,
+        git.rewrite_commit_messages(prepared_commits.as_mut_slice(), None),
+    );
+
+    result
+}
+
+async fn close_impl(
+    gh: &mut crate::github::GitHub,
+    config: &crate::config::Config,
+    prepared_commit: &mut PreparedCommit,
+) -> Result<()> {
+    let pull_request_number =
+        if let Some(number) = prepared_commit.pull_request_number {
+            output("#️⃣ ", &format!("Pull Request #{}", number))?;
+            number
+        } else {
+            return Err(Error::new(
+                "This commit does not refer to a Pull Request.",
+            ));
+        };
+
+    // Load Pull Request information
+    let pull_request = gh.get_pull_request(pull_request_number).await?;
+
+    if pull_request.state != PullRequestState::Open {
+        return Err(Error::new(formatdoc!(
+            "This Pull Request is already closed!",
+        )));
+    }
+
+    output("📖", "Getting started...")?;
+
+    let base_is_master = pull_request.base.is_master_branch();
+
+    let result = gh
+        .update_pull_request(
+            pull_request_number,
+            PullRequestUpdate {
+                state: Some(PullRequestState::Closed),
+                ..Default::default()
+            },
+        )
+        .await;
+
+    match result {
+        Ok(closed) => closed,
+        Err(error) => {
+            output("❌", "GitHub Pull Request close failed")?;
+
+            return Err(error);
+        }
+    };
```

Comment:
As the reader of this code, I now wonder what the type of `closed` is, and if we should do anything with it. We pass it out of the `Ok`-branch of this `match`, but then it's discarded.

I checked out the branch and saw that `closed` is `()`.

I suggest to change the `Ok` branch like this: `Ok(()) => (),`, then it's self explanatory.

My initial reaction was to think that I'd like `if let Err(error) = result` more, but now thinking about it, that also just ignores whatever non-error value `result` could have. So now I like using `match` here. If you replace `closed` in there with `()`, then the code makes it clear on the first sight: `result` can be one of two things, either an empty Ok (`Ok(())`), or an error.

I am still very much in the learning phase of Rust. This comment is me thinking out loud.

Reply:
This all makes sense!

I tell you, as someone with more experience in purely functional languages (PureScript, Haskell, elm, etc.) I expected that if this didn't evaluate to `()` it would be either a compile error or at least a lint error for an unused value. Similar to how if you were to write like `123;` on a line all by itself I would also expect it to be a compile/lint error. I see now that that's not the case, and that's probably fine. Just not what I expected.

I think what you suggest makes sense, though. Lemme update this.

Reply:
There is a way in Rust to mark functions such that its return value must not be ignored. Or maybe it was types you mark. I did notice that you can't just ignore `Result<_, _>`, because that would mean to silently ignore errors potentially, which shouldn't happen. Don't remember from the top of my head how it's done, maybe some `#[]`-style directive.

But by default, Rust doesn't complain. I think I'm okay with this approach. Only forces you to write boilerplate to explicitly ignore something, if that something normally shouldn't be ignored.